prov/verbs: Workaround for #4672 and IBV_WC_WR_FLUSH_ERR WCs

sydidelot opened the pull request:
Since the patch fe16dd1("prov/verbs: Fix a bug during connection
shutdown from remote side") that attempts to fix an issue with
IWarp, the Verbs provider silently skips every WCs failed with
the status IBV_WC_WR_FLUSH_ERR.
This causes the end-user application to never receive these WCs
for network protocols other than IWarp, which is a regression
since Libfabric 1.4.0.

This patch works around the problem and add a selective code path
if the CQ is for an IWarp device.

Signed-off-by: Sylvain Didelot <sdidelot@ddn.com>

dmitrygladkov commented:
Ops, sorry I missed this issue.

Could you please help me to understand this in more details?
Does your OFI based application expect to receive the WCs with status `IBV_WC_WR_FLUSH_ERR`?
If yes -- do you want to receive them when EP that is bound to the CQ is still operational (before `fi_close` was called for this EP)? or after EP was closed?

Thanks!

sydidelot commented:
> Ops, sorry I missed this issue.

No worry, @dmitrygladkov! I appreciate your help on this issue!

Our OFI based application expects to receive a send work completion for every network message successfully queued to Libfabric.

Internally, our application keeps a reference counter that tracks the number of TX outstanding messages which are not yet completed for a specific connection.

When the application receives a disconnect event for a connection (FI_SHUTDOWN), it waits for all the TX oustanding messages to be canceled by the network (via the IBV_WC_WR_FLUSH_ERR WR status). The application finally calls fi_close() to destroy the endpoint only when the last references is dropped.

After fi_close() is called, we don't expect any further events from the CQ for this specific endpoint.

Does it answer your question or do you want me to elaborate more about our application?
Thanks very much for your help.

Sylvain

dmitrygladkov commented:
> Does it answer your question or do you want me to elaborate more about our application?

@sydidelot  Yes, it answers my question with needed level of details! Thank you! 

I think we should just remove the code added by me in that PR commit -- i.e. remove discarding of `IBV_WC_WR_FLUSH_ERR WR`.
RxM provider must handle this as your application does.

Let me a couple of days to take a look at this bug and provide the fix for verbs/MSG and RxM/Verbs.

sydidelot commented:
> > Does it answer your question or do you want me to elaborate more about our application?
> 
> @sydidelot Yes, it answers my question with the needed level of details! Thank you!
> 
> I think we should just remove the code added by me in that PR commit -- i.e. remove discarding of `IBV_WC_WR_FLUSH_ERR WR`.
> RxM provider must handle this as your application does.
> 
> Let me a couple of days to take a look at this bug and provide the fix for verbs/MSG and RxM/Verbs.

Thanks a lot! Let me know if there is anything I can do to help you!

sydidelot commented:
@dmitrygladkov Have you had the chance to look at the issue #4672 ?
I have updated the pull request with a version where I removed all the specific code path related to `IBV_WC_WR_FLUSH_ERR`. Let me know if you prefer this version of the patch.

Thank you!

shefty commented:
Dmitry needed to move away from Intel. Asking @a-ilango to look at these changes.

dmitrygladkov commented:
@sydidelot Yes, this makes sense to me. But we need to be sure that this works fine with RxM/Verbs (RxM posts `fi_recv()`s for implementing bounce buffers). When we do fi_close() for a specific Verbs EP from RxM CM, Verbs should complete all posted WQEs during fi_close(). 

> I hit a condition where the Verbs CQ is busy and can not be destroyed.

@swelch yes, this is known issue for me

a-ilango commented:
As per the man page, fi_shutdown has the following:

"Outstanding operations posted to the endpoint when fi_shutdown is called will be canceled or discarded. "

I think it's similar behavior for remote shutdown though the man page doesn't seem to explicitly state it.

So it's provider's choice to discard (silently) or cancel (generate FI_CANCELED completion errors I think). Is it preferred that the provider cancel operations instead of silently discard them?

>  on a rare occasion at termination I hit a condition where the Verbs CQ is busy and can not be destroyed.

I wonder if it's because the verbs completion queue has pending work completions to be read.

shefty commented:
The ability to return canceled operation is provider specific.  The man pages reflect this.  However, the behavior of the verbs provider changed, which is what @sydidelot would like to revert.

And it's a @dmitrygladkov !  :)

a-ilango commented:
Okay. We can preserve the behavior that was in libfabric 1.4. I think RxM needs to keep a reference count similar to what @sydidelot 's app does. I'll prepare a fix.

sydidelot commented:
@a-ilango Thank you!

a-ilango commented:
@sydidelot btw, does only iWARP throw these FLUSH errors for incomplete requests when the QP is closed? I don't see the same behavior for Infiniband QPs.

sydidelot commented:
@a-ilango My original issue was for Infiniband QPs (Mellanox HCAs). I haven't tested with iWARP.
If you close an Infiniband QP that has outstanding TX messages -- or the remote process decides to close the connection --, you will likely get WCs events generated with status `IBV_WC_WR_FLUSH_ERR`.

a-ilango commented:
I see. I was just having outstanding receives when the verbs QP is closed.

a-ilango commented:
RxM calls fi_close on the verbs endpoints when FI_SHUTDOWN is received or if rxm_endpoint is closed. It doesn't call fi_shutdown anywhere. The man page says completions would be discarded on fi_close.

On fi_close for a MSG endpoint, verbs provider polls any outstanding CQ errors and enqueues them to verbs provider software CQ and those completion entries /errors would be discarded on CQ close.

So I think RxM wouldn't need any change here. This patch can be added to verbs provider.

shefty commented:
I'm assuming that PR #4848 will replace this one and allow it to be closed once merged.  Marking as do no merge as a result.

a-ilango commented:
Closing this as #4848 got merged.
